sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamTableFunctionScanRel.java

```diff
@@ -0,0 +1,162 @@
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.plan.RelTraitSet;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rel.RelNode;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rel.core.TableFunctionScan;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rel.metadata.RelColumnMapping;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rel.metadata.RelMetadataQuery;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rel.type.RelDataType;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rex.RexCall;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rex.RexInputRef;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rex.RexLiteral;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rex.RexNode;
+import org.joda.time.Duration;
+
+/**
+ * BeamRelNode to replace {@code TableFunctionScan}. Currently this class limits to support
+ * table-valued function for streaming windowing.
+ */
+public class BeamTableFunctionScanRel extends TableFunctionScan implements BeamRelNode {
+  public BeamTableFunctionScanRel(
+      RelOptCluster cluster,
+      RelTraitSet traitSet,
+      List<RelNode> inputs,
+      RexNode rexCall,
+      Type elementType,
+      RelDataType rowType,
+      Set<RelColumnMapping> columnMappings) {
+    super(cluster, traitSet, inputs, rexCall, elementType, rowType, columnMappings);
+  }
+
+  @Override
+  public TableFunctionScan copy(
+      RelTraitSet traitSet,
+      List<RelNode> list,
+      RexNode rexNode,
+      Type type,
+      RelDataType relDataType,
+      Set<RelColumnMapping> set) {
+    return new BeamTableFunctionScanRel(
+        getCluster(), traitSet, list, rexNode, type, relDataType, columnMappings);
+  }
+
+  @Override
+  public PTransform<PCollectionList<Row>, PCollection<Row>> buildPTransform() {
+    return new Transform();
+  }
+
+  private class Transform extends PTransform<PCollectionList<Row>, PCollection<Row>> {
+
+    @Override
+    public PCollection<Row> expand(PCollectionList<Row> input) {
+      checkArgument(
+          input.size() == 1,
+          "Wrong number of inputs for %s, expected 1 input but received: %s",
+          BeamTableFunctionScanRel.class.getSimpleName(),
+          input);
+      String operatorName = ((RexCall) getCall()).getOperator().getName();
+      checkArgument(
+          operatorName.equals("TUMBLE"),
+          "Only support TUMBLE table-valued function. Current operator: %s",
+          operatorName);
+      RexCall call = ((RexCall) getCall());
+      RexInputRef wmCol = (RexInputRef) call.getOperands().get(1);
+      PCollection<Row> upstream = input.get(0);
+      Schema outputSchema = CalciteUtils.toSchema(getRowType());
+      return upstream
+          .apply(
+              ParDo.of(
+                  new FixedWindowDoFn(
```

Comment:
This still appears to actually be a "FixedWindowsTableFunctionRel".

Reply:
Interesting point. Let me check if there are benefits to split TUMBLE/HOP/SESSION at Rel level, especially what is the impact on JOIN.

Reply:
Ok, I have thought about this. Can we keep what this code is since it will still fast evolving. 

There are a few dimensions to consider:
1. if TUMBLE, SESSION, HOP have their own Rels
2. How many Rels needed for JOIN with windowing 


I believe we will have better answer after we see HOP and SESSION implementation.

Reply:
My point is really that committing the code like this makes code with names that don't make sense. If you get interrupted to do something else after this PR, then the codebase will be in an incorrect state.

Reply:
I guess I got what you meant: you think it should be a PTransform, not a DoFn for this `FixedWindow` assignment? (I was confused on why this transforming should be a Rel)

If so, either PTransform or a DoFn seems work to me.